crates/bevy_ui/src/flex/mod.rs

```diff
@@ -37,257 +40,376 @@ impl Default for FlexSurface {
+        flex_surface
+            .stretch
+            .compute_layout(*window_node, stretch::geometry::Size::undefined())
+            .unwrap();
     }
+}
+
+fn update_transforms_xy(
+    flex_surface: &mut FlexSurface,
+    transform_xy_query: &mut Query<
+        (Entity, &mut Node, &mut Transform, Option<&Parent>),
+        With<Style>,
+    >,
+) {
+    for (entity, mut node, mut transform, parent) in transform_xy_query.iter_mut() {
+        let stretch_node = flex_surface.entity_to_stretch.get(&entity).unwrap();
+        let layout = flex_surface.stretch.layout(*stretch_node).unwrap();
+        node.size = Vec2::new(layout.size.width, layout.size.height);
 
-    fn update_changed<F>(
-        flex_surface: &mut FlexSurface,
-        scaling_factor: f64,
-        query: Query<(Entity, &Style, Option<&CalculatedSize>), F>,
-    ) where
-        F: QueryFilter,
-    {
-        // update changed nodes
-        for (entity, style, calculated_size) in query.iter() {
-            // TODO: remove node from old hierarchy if its root has changed
-            if let Some(calculated_size) = calculated_size {
-                flex_surface.upsert_leaf(entity, &style, *calculated_size, scaling_factor);
-            } else {
-                flex_surface.upsert_node(entity, &style, scaling_factor);
+        let position = &mut transform.translation;
+        position.x = layout.location.x + layout.size.width / 2.0;
+        position.y = layout.location.y + layout.size.height / 2.0;
+        if let Some(parent) = parent {
+            if let Some(stretch_node) = flex_surface.entity_to_stretch.get(&**parent) {
+                let parent_layout = flex_surface.stretch.layout(*stretch_node).unwrap();
+
+                position.x -= parent_layout.size.width / 2.0;
+                position.y -= parent_layout.size.height / 2.0;
             }
         }
     }
+}
 
-    for (entity, style, calculated_size) in changed_size_query.iter() {
-        flex_surface.upsert_leaf(entity, &style, *calculated_size, logical_to_physical_factor);
-    }
-
-    // TODO: handle removed nodes
-
-    // update window children (for now assuming all Nodes live in the primary window)
-    if let Some(primary_window) = windows.get_primary() {
-        flex_surface.set_window_children(primary_window.id(), root_node_query.iter());
+pub fn z_index_system(
+    changed_zindex_query: Query<
+        Entity,
+        (
+            With<Style>,
+            With<Transform>,
+            Or<(Changed<Style>, Changed<Parent>, Changed<Children>)>,
+        ),
+    >,
+    mut transform_z_query: Query<(&Style, Option<&Children>, Option<&Parent>, &mut Transform)>,
+) {
+    for entity in changed_zindex_query.iter() {
+        update_one_transform_z(entity, &mut transform_z_query);
     }
+}
 
-    // update children
-    for (entity, children) in children_query.iter() {
-        flex_surface.update_children(entity, &children);
+fn update_one_transform_z(
```

Comment:
This does a ton of redunant work in terms of allocations, traversals, sorting, etc (especially on startup). I think it makes sense to add some tracking here to cut down on the combinatorics.

Ex: On startup _every single_ node:

1. walks up the tree to (worst case) the root O(logn)
2. Allocates/populates a stacking context with every node (worst case the whole tree, which is O(n))
3. Sorts the stacking context O(nlogn)
4. Sets the transform z value of each node in the stacking context (which is O(n))

By my calculations thats worst case O(n^2logn) compute AND allocations, which is not fantastic

Reply:
>Ex: On startup every single node:
> 1. walks up the tree to (worst case) the root O(logn)

This is a real concern but I'd like to point out that it only does that work for their stacking context, which by default and in most cases (`ZIndex::None` or `ZIndex::Some(i)`) means their children. The worst case happens if the user explicitly selects `ZIndex::Auto` in every single node. In practice I only expect that feature to be used inside the small hierarchy of a widget for instance.

>3. Sorts the stacking context O(nlogn)

In most cases (`ZIndex::None` and `ZIndex::Auto`), the stacking context should already be sorted at creation (I need to test that). The sorting only has to move nodes with `ZIndex::Some(i)`.

Here is a list of potential improvements:

- use a smallvec so that most cases without `ZIndex::Auto` do not allocate
- skip sorting if there is no `ZIndex::Some(i)`
- cache the stacking context (maybe only when `ZIndex::Auto` is involved?)
- use an iterator to add all children at once instead of pushing one by one
- as discussed, making ZIndex a separate component so that changes to other Style properties do not trigger this whole process. (Alternatively, ZIndex could have its own mutation flag?)

The part that I'm unsatisfied with is that the process is duplicated when several non leaf entities from the same stacking context are added/updated at once. However I'm being cautious because it would be unfortunate if an optimisation for that adds overhead for the default/common cases. I'll think about what kind of tracking can be done there.

Besides, I think there is a bug in this implementation for negative z-index, I'll have to check.

Last thought: is the `ZIndex::Auto` feature worth adding all that complexity?

Reply:
The new commit only allocates stacking contexts once, and makes sure they are updated at most once every frame, when required.
However the z transforms are now all updated if any stacking context changed, since using the transforms's scale for incremental update was causing issues with f32 precision.